src/lib/IDBDAO.js

```diff
@@ -0,0 +1,377 @@
+/**
+ * @license
+ * Copyright 2014 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'IDBException',
+  extends: 'foam.dao.ExternalException',
+
+  properties: [
+    'id',
+    'error',
+    {
+      name: 'message',
+      expression: function(id, error) {
+        return "IndexedDB Error for " + id +
+          ( error ? ": " + error.toString() : "" );
+      }
+    }
+  ]
+});
+
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'IDBDAO',
+  extends: 'foam.dao.AbstractDAO',
+
+  requires: [
+    'foam.dao.FlowControl',
+    'foam.dao.ArraySink',
+    'foam.dao.IDBException',
+  ],
+
+  imports: [
+    'async',
+  ],
+
+//   function documentation() {/*
+//   Usage:<br/>
+//    <code>var dao = IDBDAO.create({model: Issue});<br/>
+//    var dao = IDBDAO.create({model: Issue, name: 'ImportantIssues'});<br/></code>
+//   <br/>
+//    TODO:<br/>
+//    Optimization.  This DAO doesn't use any indexes in indexeddb yet, which
```

Comment:
I don't know if this is even an TODO worth having.  Optimizing against IDB is a losing battle.  Maybe some minor improvements can be made but in general you're not going to get very far, better off caching in an MDAO.